src/openrct2/actions/RideCreateAction.hpp

```diff
@@ -0,0 +1,314 @@
+* (at your option) any later version.
+*
+* A full copy of the GNU General Public License can be found in licence.txt
+*****************************************************************************/
+#pragma endregion
+
+#pragma once
+
+#include "../core/Memory.hpp"
+#include "../core/MemoryStream.h"
+#include "../localisation/string_ids.h"
+#include "GameAction.h"
+
+extern "C"
+{
+    #include "../localisation/date.h"
+    #include "../cheats.h"
+    #include "../interface/window.h"
+    #include "../world/park.h"
+    #include "../rct1.h"
+    #include "../ride/ride_data.h"
+    #include "../ride/ride.h"
+    #include "../ride/station.h"
+}
+
+struct RideCreateGameActionResult : public GameActionResult
+{
+    RideCreateGameActionResult() : GameActionResult(GA_ERROR::OK, 0) {}
+    RideCreateGameActionResult(GA_ERROR error, rct_string_id message) : GameActionResult(error, message) {}
+
+    sint32 rideIndex;
+    uint32 rideColor;
+};
+
+struct RideCreateAction : public GameActionBase<GAME_COMMAND_CREATE_RIDE, GA_FLAGS::ALLOW_WHILE_PAUSED, RideCreateGameActionResult>
+{
+public:
+    sint32 rideType;
+    sint32 rideSubType;
+
+    void Serialise(DataSerialiser& stream) override
+    {
+        GameAction::Serialise(stream);
+
+        stream << rideType << rideSubType;
+    }
+
+    GameActionResult::Ptr Query() const override
+    {
+        if (rideType >= RIDE_TYPE_COUNT)
+        {
+            return std::make_unique<RideCreateGameActionResult>(GA_ERROR::INVALID_PARAMETERS, STR_INVALID_RIDE_TYPE);
+        }
+
+        return std::make_unique<RideCreateGameActionResult>();
+    }
+
+    GameActionResult::Ptr Execute() const override
+    {
+        rct_ride_entry * rideEntry;
+        auto res = std::make_unique<RideCreateGameActionResult>();
+
+        sint32 subType = rideSubType;
+        if (subType == RIDE_ENTRY_INDEX_NULL)
+        {
+            uint8 *availableRideEntries = get_ride_entry_indices_for_ride_type(rideType);
+            for (uint8 *rei = availableRideEntries; *rei != RIDE_ENTRY_INDEX_NULL; rei++)
+            {
+                rideEntry = get_ride_entry(*rei);
+
+                // Can happen in select-by-track-type mode
+                if (!ride_entry_is_invented(*rei) && !gCheatsIgnoreResearchStatus)
+                {
+                    continue;
+                }
+
+                if (!(rideEntry->flags & RIDE_ENTRY_FLAG_SEPARATE_RIDE_NAME) || rideTypeShouldLoseSeparateFlag(rideEntry))
+                {
+                    subType = *rei;
+                    break;
+                }
+            }
+            if (subType == RIDE_ENTRY_INDEX_NULL)
+            {
+                subType = availableRideEntries[0];
+                if (subType == RIDE_ENTRY_INDEX_NULL)
+                {
+                    res->Error = GA_ERROR::INVALID_PARAMETERS;
+                    res->ErrorMessage = STR_INVALID_RIDE_TYPE;
+                    return std::move(res);
```

Comment:
This should be checked in the `Query`. There is no reason to send a command to the server when even the client cannot complete it.